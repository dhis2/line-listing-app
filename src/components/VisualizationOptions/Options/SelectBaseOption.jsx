import { SingleSelectField, SingleSelectOption } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React from 'react'
import { connect } from 'react-redux'
import { acSetUiOptions } from '../../../actions/ui.js'
import { sGetUiOptions } from '../../../reducers/ui.js'
import styles from '../styles/VisualizationOptions.module.css'

const SelectBaseOption = ({ option, label, value, onChange, dataTest }) => {
    const selected = option.items.find((item) => item.value === value)?.value

    return (
        <div className={styles.tabSectionOption}>
            <SingleSelectField
                name={`${option.name}-select`}
                label={label}
                onChange={({ selected }) => onChange(selected)}
                selected={selected}
                inputWidth="280px"
                dense
                dataTest={`${dataTest}-select`}
            >
                {option.items.map(({ value, label }) => (
                    <SingleSelectOption
                        key={value}
                        value={value}
                        label={label}
                        dataTest={`${dataTest}-option`}
                    />
                ))}
            </SingleSelectField>
        </div>
    )
}

SelectBaseOption.propTypes = {
    option: PropTypes.object.isRequired,
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    onChange: PropTypes.func.isRequired,
    dataTest: PropTypes.string,
    label: PropTypes.string,
}

const mapStateToProps = (state, ownProps) => ({
    value: sGetUiOptions(state)[ownProps.option.name],
})

const mapDispatchToProps = (dispatch, ownProps) => ({
    onChange: (value) =>
        dispatch(acSetUiOptions({ [ownProps.option.name]: value })),
})

export default connect(mapStateToProps, mapDispatchToProps)(SelectBaseOption)
