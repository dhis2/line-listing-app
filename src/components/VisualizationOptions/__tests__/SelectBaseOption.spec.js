import { render, screen } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import SelectBaseOption from '../Options/SelectBaseOption.js'

const mockStore = configureMockStore()

describe('ER > Options > SelectBaseOption', () => {
    it('renders the list of options', () => {
        const store = {
            ui: {
                options: {
                    theOptionName: 'option3',
                },
            },
        }

        const option = {
            name: 'theOptionName',
            items: [
                { label: 'Option 1 label', value: 'option1' },
                { label: 'Option 2 label', value: 'option2' },
                { label: 'Option 3 label', value: 'option3' },
            ],
        }

        const { getByText } = render(
            <Provider store={mockStore(store)}>
                <SelectBaseOption
                    dataTest="testing-prefix"
                    label="Should I?"
                    option={option}
                />
            </Provider>
        )

        expect(getByText('Option 3 label')).toBeTruthy()

        const option2 = screen.queryByText('Option 2 label')
        expect(option2).toBeNull()

        const option1 = screen.queryByText('Option 1 label')
        expect(option1).toBeNull()
    })
})
