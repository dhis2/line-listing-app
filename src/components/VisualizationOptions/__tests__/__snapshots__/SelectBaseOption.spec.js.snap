// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ER > Options > SelectBaseOption renders the list of options 1`] = `
<div>
  <div
    class="tabSectionOption"
  >
    <div
      class="jsx-1141087211 "
      data-test="testing-prefix-select"
    >
      <label
        class="jsx-2718078600 "
        data-test="testing-prefix-select-label"
      >
        <span
          class="jsx-2718078600"
        >
          Should I?
        </span>
      </label>
      <div
        class="jsx-3111771310 "
        data-test="testing-prefix-select-content"
      >
        <div
          class="jsx-15480391 "
          data-test="dhis2-uicore-box"
        >
          <div
            class="jsx-114080822 root"
            data-test="dhis2-uicore-singleselect"
          >
            <div
              class="jsx-114080822 root-input"
            >
              <div
                data-test="dhis2-uicore-select"
              >
                <div
                  class="jsx-1165050001 root dense"
                  data-test="dhis2-uicore-select-input"
                  tabindex="0"
                >
                  <div
                    class="jsx-1165050001 root-children"
                  >
                    <div
                      class="jsx-2240447017 jsx-4123978364 root"
                    >
                      <div
                        class="jsx-2240447017 jsx-4123978364 root-input"
                      >
                        <div
                          class="jsx-1323585369 root"
                        >
                          Option 3 label
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="jsx-1165050001 root-right"
                  >
                    <svg
                      height="16"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M4.146 6.146a.5.5 0 01.638-.057l.07.057L8 9.293l3.146-3.147a.5.5 0 01.638-.057l.07.057a.5.5 0 01.057.638l-.057.07-3.5 3.5a.5.5 0 01-.638.057l-.07-.057-3.5-3.5a.5.5 0 010-.708z"
                        fill="currentColor"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
